### ISSUE TITLE
UART interrupt pin state should be handled for migration

### ISSUE THREAD

**Issue by jordanhendricks** [oxidecomputer/propolis#390]
The 16550 UART emulation has a field called `intr_pin` that represents whether the device has a pending interrupt.

My understanding of the chain of events here from setting this field, to generating an interrupt in the guest is:
- the 16550 UART sets `intr_pin` to true
- the LPC UART calls `sync_intr_pin` any time it might have caused mutation in the UART state
- `sync_intr_pin` calls into `LegacyPIC`, asserting the interrupt if `intr_pin` is true, and deasserting it otherwise
- the assert/deasserts become `VM_ISA_{ASSERT,DEASSERT}_IRQ` ioctls to bhyve
- the bhyve vatpic handles these ioctls, updating its internal state of the interrupt pins and changing the level of pins on the lapic and iopic as needed
- interrupts on the lapic are dispensed to the guest via changing some data in the vm control structure

The `intr_pin` field is currently not migrated. As it is today, I believe this means the guest could lose an interrupt. In any event, it will likely be in an inconsistent state and should be handled.

**Comment by jordanhendricks** [comment]
A key observation here is that changes to state in the UART ultimately get pushed to bhyve, which can update the state of one or more other devices in-kernel (namely, the vatpic, lapic, viopic). Since we have no strict guarantees about the order of device migration, we could potentially migrate the PIC state before migrating devices that use them (like the uart). If the devices are not paused (as is also true of the UART), these devices may continue to change in-kernel state even after we have migrated the underlying kernel devices.

To have a consistent device snapshot here, I think we also need make sure that any devices using the PIC are paused before we migrate its state. It seems we have a dependency graph here of device state, so perhaps we could have a more first-order concept of device state dependencies and pause them in reverse order?

Some helpful context from vatpic.c about how PIC interrupts are routed, which illustrates which in-kernel devices might have state change from PIC interrupts:
```
                /*                                                                                                                                                                                                                                                            
                 * From Section 3.6.2, "Interrupt Modes", in the                                                                                                                                                                                                              
                 * MPtable Specification, Version 1.4                                                                                                                                                                                                                         
                 *                                                                                                                                                                                                                                                            
                 * PIC interrupts are routed to both the Local APIC                                                                                                                                                                                                           
                 * and the I/O APIC to support operation in 1 of 3                                                                                                                                                                                                            
                 * modes.                                                                                                                                                                                                                                                     
                 *                                                                                                                                                                                                                                                            
                 * 1. Legacy PIC Mode: the PIC effectively bypasses                                                                                                                                                                                                           
                 * all APIC components.  In this mode the local APIC is                                                                                                                                                                                                       
                 * disabled and LINT0 is reconfigured as INTR to                                                                                                                                                                                                              
                 * deliver the PIC interrupt directly to the CPU.                                                                                                                                                                                                             
                 *                                                                                                                                                                                                                                                            
                 * 2. Virtual Wire Mode: the APIC is treated as a                                                                                                                                                                                                             
                 * virtual wire which delivers interrupts from the PIC                                                                                                                                                                                                        
                 * to the CPU.  In this mode LINT0 is programmed as                                                                                                                                                                                                           
                 * ExtINT to indicate that the PIC is the source of                                                                                                                                                                                                           
                 * the interrupt.                                                                                                                                                                                                                                             
                 *                                                                                                                                                                                                                                                            
                 * 3. Virtual Wire Mode via I/O APIC: PIC interrupts are                                                                                                                                                                                                      
                 * fielded by the I/O APIC and delivered to the appropriate                                                                                                                                                                                                   
                 * CPU.  In this mode the I/O APIC input 0 is programmed                                                                                                                                                                                                      
                 * as ExtINT to indicate that the PIC is the source of the                                                                                                                                                                                                    
                 * interrupt.                                                                                                                                                                                                                                                 
                 */   
```

**Comment by pfmooney** [comment]
>  I think we also need make sure that any devices using the PIC are paused before we migrate its state

We need to make sure _all_ devices are paused before migrating state.  Those using MSI will still incur changes to the kernel-emulated devices if they continue to operate while we're gathering the migration payload.

> Since we have no strict guarantees about the order of device migration, we could potentially migrate the PIC state before migrating devices that use them (like the uart). 

There are already some assumptions baked in about how we migrate the state of the pin-based interrupt controllers (AT-PIC and IO-APIC).  The in-kernel logic has the "canonical" state of those pins, as exported/import as part of the migration.  Parallel to that, the propolis userspace has state about the devices emulated there (uarts, PCI lintrs, etc) and if they are asserted or not.  If any of those userspace pins are asserted at the time of migration, they'll need to be restored to that state on the destination side - a process which would typically involve ioctls in order for the state to be reflected by the in-kernel-emulated devices.  Since those kernel devices have already been migrated with that asserted state, those updates from userspace should be elided during importation.

In short, I believe that we'll need some sort of addition to the existing interface(s) which allows the userspace portions of the ATPIC/IOAPIC emulation to ignore state changes to their subordinate pin resources in userspace while importing device data.  After all devices have been imported, the pins asserted from those userspace devices should match the state loaded by the in-kernel emulation side of the PICs

**Comment by hawkw** [comment]
If I understand correctly, the core issue here is as follows:
1. Both Bhyve (in the kernel) and Propolis (in userspace) track the state of PIC and IOAPIC interrupt pins. In particular, both Propolis and the kernel count the number of times a pin has been asserted. In Propolis, we use this level count to determine whether we actually perform the `ioctl` call that tells the kernel to change the pin's state, so that we assert the pin only when going from 0->1 and deassert it only when going from 1->0.
2. When migrating a VM, we treat the kernel's state as the single source of truth when we import the `BhyveAtPic` and `BhyveIoApic` devices.
3. The userspace state of the ATPIC is tracked in a separate abstraction, `LegacyPIC`, which is not imported/exported during migration. This is what the UART uses to actually send an IRQ to the guest. Pin level counts on `LegacyPIC` are used to determine whether to call the `ioctl` that actually asserts/deasserts a pin in the guest VM.
4. During migration, we don't guarantee the order in which devices are imported or exported. We may import the `BhyveAtPic` device before we import the UART. If this occurs, we will set the pin level counts in Bhyve to match those exported from the source VM. However, when we import the UART, it may have an interrupt pending, and may attempt to assert the UART interrupt pin using `LegacyPIC`...and the `LegacyPIC` doesn't realize the kernel thinks the pin is already asserted, causing the guest to see it asserted a second time.

If that's all correct, it seems like one potential solution would be to ensure that, when migrating the chipset (which owns `LegacyPIC`), we read the PIC IRQ pin states from the kernel, and set the counts for the userspace-tracked IRQ pins to match the kernel-tracked IRQ pins.

An alternative approach could be actually migrating the userspace-tracked pin states in addition to the kernel-tracked pin states. However, I think it's probably nicer to just have a single source of truth in the kernel, and ensure that the userspace state matches it. 

@pfmooney, does that seem like a sufficient solution for this, or am I overlooking something?

**Comment by pfmooney** [comment]
>     1. Both Bhyve (in the kernel) and Propolis (in userspace) track the state of PIC and IOAPIC interrupt pins. In particular, both Propolis and the kernel count the number of times a pin has been asserted. In Propolis, we use this level count to determine whether we actually perform the `ioctl` call that tells the kernel to change the pin's state, so that we assert the pin only when going from 0->1 and deassert it only when going from 1->0.

Yeah, Propolis keeps its own counts for pins in order to decide when to make assert/deassert calls into the kernel.  Because the kernel keeps its own counts, it could be that an assert or deassert from userspace will not result in an interrupt delivery to the guest due to existing pin level from a kernel device (the vHPET is the only kernel-emulated device which can be configured for level-triggered operations on its pins).

>     2. When migrating a VM, we treat the kernel's state as the single source of truth when we import the `BhyveAtPic` and `BhyveIoApic` devices.

Considering `BhyveAtPic` and `BhyveIoApic` are the kernel representations of those devices, yes.

>     3. The userspace state of the ATPIC is tracked in a separate abstraction, `LegacyPIC`

Not entirely accurate.  The assert/deassert calls from LegacyPIC impact both the ATPIC and IOAPIC devices.

> which is not imported/exported during migration

Correct that do not (today) export the pin assert counts in `LegacyPIC`.


>  This is what the UART uses to actually send an IRQ to the guest. Pin level counts on `LegacyPIC` are used to determine whether to call the `ioctl` that actually asserts/deasserts a pin in the guest VM.

Whether or not the `ioctl` call is made to change the in-kernel counts (which may cause interrupt delivery).

>     4. During migration, we don't guarantee the order in which devices are imported or exported. We may import the `BhyveAtPic` device before we import the UART. If this occurs, we will set the pin level counts in Bhyve to match those exported from the source VM. However, when we import the UART, it may have an interrupt pending, and may attempt to assert the UART interrupt pin using `LegacyPIC`...and the `LegacyPIC` doesn't realize the kernel thinks the pin is already asserted, causing the guest to see it asserted a second time.

Yeah, it's for that reason that the UART _isn't_ properly setting its pin state today - to avoid extra interrupt assertions.

> If that's all correct, it seems like one potential solution would be to ensure that, when migrating the chipset (which owns `LegacyPIC`), we read the PIC IRQ pin states from the kernel, and set the counts for the userspace-tracked IRQ pins to match the kernel-tracked IRQ pins.

This is a bit tricky.  Firstly userspace is not the only consumer which manipulates those counts.  As noted above, the HPET can also assert pins in a level-triggered manner.  Beyond that, we need to be very clear what it means to "set the counts" in userspace.  We could set the pin assertion level in the LegacyPIC, yes, but if other devices which we are importing have interrupt pins which are high, we would need to elide those assertions to avoid double-counting them.

> An alternative approach could be actually migrating the userspace-tracked pin states in addition to the kernel-tracked pin states. However, I think it's probably nicer to just have a single source of truth in the kernel, and ensure that the userspace state matches it.

I agree that it would be nice to keep the migration state for a device concentrated in a single place.  How we choose to do that is an open question.

**Comment by hawkw** [comment]
@pfmooney  and I spoke about this a bit more, and we determined that we probably ought to migrate the userspace-generated interrupt pin states separately from kernelspace, because there might be additional pin assertions that are generated from kernel-emulated devices such as the HPET. Patrick proposed the idea of adding new methods to the `IntrPins` trait that assert/deassert pins _during a migration_, which could behave differently from the methods called outside of migrations.

Also, we agreed that it probably makes sense to base any changes to the interrupt pin code on #639 to minimize merge conflicts.
